extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,156 @@
+
+#include "extensions/stackdriver/common/constants.h"
+#include "extensions/stackdriver/edges/edges.pb.h"
+#include "google/protobuf/util/time_util.h"
+
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+#else
+#include "extensions/common/wasm/null/null_plugin.h"
+#endif
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::cloud::meshtelemetry::v1alpha1::TrafficAssertion;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_GRPC;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTP;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTPS;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_TCP;
+using google::cloud::meshtelemetry::v1alpha1::WorkloadInstance;
+using google::protobuf::util::TimeUtil;
+
+void instanceFromMetadata(const ::wasm::common::NodeInfo& node_info,
+                          WorkloadInstance* instance) {
+  // TODO(douglas-reid): support more than just kubernetes instances
+  instance->set_uid("kubernetes://" + node_info.name() + "." +
+                    node_info.namespace_());
+  // TODO(douglas-reid): support more than just GCP ?
+  instance->set_location(
+      node_info.platform_metadata().at(Common::kGCPClusterLocationKey));
+  instance->set_cluster_name(
+      node_info.platform_metadata().at(Common::kGCPClusterNameKey));
+  instance->set_owner_uid(node_info.owner());
+  instance->set_workload_name(node_info.workload_name());
+  instance->set_workload_namespace(node_info.namespace_());
+  return;
+};
+
+EdgeReporter::EdgeReporter(
+    const ::wasm::common::NodeInfo& local_node_info,
+    std::unique_ptr<MeshEdgesServiceClient> edges_client) {
+  current_request_ = std::make_unique<ReportTrafficAssertionsRequest>();
+
+  const auto& project_id =
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  std::string mesh_uid = "unknown";
+  current_request_->set_mesh_uid(
+      "//cloudresourcemanager.googleapis.com/projects/" + project_id + "/" +
+      location + "/meshes/" + cluster);
+
+  node_instance_ = std::make_unique<WorkloadInstance>();
+  instanceFromMetadata(local_node_info, node_instance_.get());
+
+  edges_client_ = std::move(edges_client);
+};
+
+std::string nodeKey(const ::wasm::common::NodeInfo& node_info) {
+  return node_info.name() + "." + node_info.namespace_();
```

Comment:
This is allocating a string every time. Suggest to save the string by value in the class.

Reply:
I added this bit to `context.cc`, as NodeInfo is a proto-generated class. I'm not sure if that is better or not.